<a href="https://colab.research.google.com/github/nilsonprof/robo_aj/blob/master/BIG_A1_LinearRegressionStatsmodels.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<head>
  <meta name="author" content="Rogério de Oliveira">
  <meta institution="author" content="Universidade Presbiteriana Mackenzie">
</head>

<img src="http://meusite.mackenzie.br/rogerio/mackenzie_logo/UPM.2_horizontal_vermelho.jpg" width=300, align="right">
<!-- <h1 align=left><font size = 6, style="color:rgb(200,0,0)"> optional title </font></h1> -->


# **Tarefa de Aprofundamento 1** 
## CASO: Estimando o Preço de Imóveis em São Paulo
---
<br>
<br>





A partir deste notebook crie um modelo de regressão múltipla com a biblioteca `statsmodels` para predição de preço de imóveis com base no histórico de preços e características de vários imóveis de São Paulo. 

Em seguida, com base nos seus resultados, responda o **questionário do Moodle**. 

# Basic imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Exploração dos Dados

Entenda os dados disponíveis e em seguida faça as transformações necessárias nos dados. Quantidades dos dados, se existem valores nulos e valores máximos e mínimos dos atributos são algumas explorações comuns. 

(Moodle) **Q1**. Quantos bairros existem no histórico e qual o bairro mais presente nos dados?



In [ ]:
df = pd.read_excel('http://meusite.mackenzie.br/rogerio/data_load/regressao_preco_imoveis.xlsx')
df.head()

In [3]:
# seu código
##df.groupby('bairro').agg({'bairro':pd.Series.value_counts})
df.bairro.value_counts()

vila-mariana           1714
paraiso                 802
vila-nova-conceicao     789
vila-sonia              307
morumbi                 129
Name: bairro, dtype: int64

**R1** 5  Bairros
##O Bairro mais presente é "vila-mariana ->> 1714 ocorrencias"

# Preparação dos Dados

Use [statsmodels.formula.api](https://www.statsmodels.org/dev/generated/statsmodels.formula.api.ols.html)

```
from statsmodels.formula.api import ols   
```

Se você empregar esta forma do `statsmodels` ele fará de modo implítico o hot encode dos dados e não há necessidade de nenhuma transformação dos dados (não faremos a normalização aqui. Embora sempre recomendável na maior parte dos casos, aqui o resultado de predição será o mesmo com o sem a normalização).

Atenção, se você empregar 
```
from statsmodels.api import OLS 
```
os resultados serão os mesmos, mas alguma preparação nos dados será necessária e a construção do modelo será um pouco diferente.

In [14]:
from statsmodels.formula.api import ols   

# nada a fazer se estiver empregando a api acima

# Estimando o modelo com o `statsmodels`

Estime o modelo.

(Moodle) **Q2**. Qual o coeficiente de determinação ajustado obtido?

Dica: empregue `result.rsquared_adj` no lugar de `result.summary()` para obter resultados com maior precisão.

(Moodle) **Q3**. Qual coeficiente estimado é menos significativo?

In [19]:
# seu código

import statsmodels.formula.api as sm 

model_1 = sm.ols (formula = "preco ~ bairro +	areaM2 + suites + dormitorios + banheiros + vagas", data=df)
result_1 = model_1.fit()

##print()
print(result_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  preco   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     1611.
Date:                Thu, 01 Sep 2022   Prob (F-statistic):               0.00
Time:                        23:33:53   Log-Likelihood:                -54290.
No. Observations:                3741   AIC:                         1.086e+05
Df Residuals:                    3731   BIC:                         1.087e+05
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
Intercept     

**R2** - Coeficiente Ajustado --> 0.795

**R3** - Coeficiente de Menor Relevância -> banheiros  (pValue-> 0.570)



# Re-estimando o modelo com o `statsmodels`

Estime novamente o modelo excluindo o atributo menos significativo (um único atributo). 

(Moodle) **Q4**. Qual o novo coeficiente de determinação ajustado obtido?


In [11]:
## Retirando a Variável "banheiros"
import statsmodels.formula.api as sm 
model_2 = sm.ols(formula = "preco ~ bairro + areaM2 + suites + dormitorios + vagas", data=df)
result_2 = model_2.fit()
print()
print(result_2.summary())


                            OLS Regression Results                            
Dep. Variable:                  preco   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     1812.
Date:                Thu, 01 Sep 2022   Prob (F-statistic):               0.00
Time:                        23:12:50   Log-Likelihood:                -54290.
No. Observations:                3741   AIC:                         1.086e+05
Df Residuals:                    3732   BIC:                         1.087e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
Intercept    

**R4** - Coeficiente: 0.795

```
# Isto está formatado como código
```




# Predição

Faça a predição do preço, com o melhor dos modelos obtidos acima, de um imóvel na Vila Mariana, com 134 m2, 1 suite, 4 dorm, 3 banheiros e 3 vagas.



(Moodle) **Q5**. Qual o preço estimado, para o melhor dos modelos obtidos, de um imóvel na Vila Mariana, com 134 m2, 1 suite, 4 dorm, 3 banheiros e 3 vagas?



In [12]:
df.columns

Index(['bairro', 'areaM2', 'suites', 'dormitorios', 'banheiros', 'vagas',
       'preco'],
      dtype='object')

In [13]:
# seu código
df2 = pd.DataFrame({'bairro':['vila-mariana'], 'areaM2':[134], 'dormitorios':4,'suites':[1], 'banheiros':[3], 'vagas':[3]})
df2.head()
result_1.predict(df2)

0    1.349653e+06
dtype: float64

**R5** O Preco de um Imóvel com as características informadas seria: 1.349.653,00 $ (Aproximadamente)